### DISCUSSION TITLE
Question: how to use custom logger with the guide 'Exit-Once'

### DISCUSSION THREAD

**Discussion by tzq0301** [uber-go/guide#189] [Q&A] [answered]
https://github.com/uber-go/guide/blob/master/style.md#exit-once

The guide recommend to use `log.Fatal(err)` to panic the error:

```go
func main() {
  if err := run(); err != nil {
    log.Fatal(err)
  }
}
```

However, should I try to init my custom logger before I call the `run()` in `main()` if I want to use my custom logger to log the error?

Or, is the action 'use my custom logger to log the error' unnecessary to do in the `main()` func?

**Comment by abhinav** [comment, marked as answer]
@tzq0301 The example uses log.Fatal because that's what most folks already use when they exit multiple times.
You have a couple options for `main`, since the purpose of that specific exit it just to print to stderr and exit:

- Use `log.Fatal`. If you want that to not print timestamps, etc. you can do:

    ```go
    func main() {
        log.SetFlags(0)
        if err := run(); err != nil {
            log.Fatal(err)
        }
    }
    ```

- Print directly to stderr:


    ```go
    func main() {
        if err := run(); err != nil {
            fmt.Fprintln(os.Stderr, err)
            os.Exit(1)
        }
    }
    ```

- Send the exit code back from `run()`,
  which uses a custom logger:

    ```go
    func main() {
        os.Exit(run())
    }

    func run() (exitCode int) {
        logger := customLogger()

        if err := stuff(); err != nil {
            logger.Error("something went wrong", err)
            return 1
        }

        return 0
    }
    ```

The guidance doesn't aim to be prescriptive about logger/program setup.

Its main thing is only that you should have exactly one place in your program
responsible for calling `os.Exit`.
